### PR TITLE
Introduce variables storage

### PR DESCRIPTION
**Pull request by fab-10**
## PR description

This is the first PR of a series to reorganize the data in db, in order to pave the way to future optimizations.
The first goal that we want to achieve is the optimization of the db for data that is static, leveraging [the BlobDB](https://github.com/facebook/rocksdb/wiki/BlobDB) feature.
For static data like blockchain or trie logs, BlobDB is a better choice since it reduces a lot the write amplification during compaction, with the result of faster initial sync and less wear out of the disk.

Now we have variable data mixed with static data in the blockchain column family, and this is not optimal, so this PR moved the _variables_ into the dedicated column family `VARIABLES` that we can further optimize for frequently changing data in future.

The PR is meant automatically migrates existing variables to the new location on startup. 

Logs during migration
```
{"@timestamp":"2023-06-06T10:14:15,156","level":"INFO","thread":"main","class":"KeyValueStoragePrefixedKeyBlockchainStorage","message":"Migrated key chainHeadHash to variables storage","throwable":""}
{"@timestamp":"2023-06-06T10:14:15,157","level":"INFO","thread":"main","class":"KeyValueStoragePrefixedKeyBlockchainStorage","message":"Migrated key finalizedBlockHash to variables storage","throwable":""}
{"@timestamp":"2023-06-06T10:14:15,158","level":"INFO","thread":"main","class":"KeyValueStoragePrefixedKeyBlockchainStorage","message":"Migrated key safeBlockHash to variables storage","throwable":""}
{"@timestamp":"2023-06-06T10:14:15,162","level":"INFO","thread":"main","class":"KeyValueStoragePrefixedKeyBlockchainStorage","message":"Migrated key forkHeads to variables storage","throwable":""}
{"@timestamp":"2023-06-06T10:14:15,163","level":"INFO","thread":"main","class":"KeyValueStoragePrefixedKeyBlockchainStorage","message":"Migrated key local-enr-seqno to variables storage","throwable":""}
```

It is possible to revert this change in case the user want to run a previous version of Besu, before downgrading the Beus, the user must run the subcommand `storage revert-variables` with the same configuration use to normally start Besu, and only after that it is possible to install a previous version of Besu.

Logs during revert
```
2023-06-06 10:15:33.201+00:00 | main | INFO  | DatabaseMetadata | Lookup database metadata file in data directory: /data/besu
2023-06-06 10:15:33.243+00:00 | main | INFO  | RocksDBKeyValueStorageFactory | Existing database detected at /data/besu. Version 2. Compacting database...
2023-06-06 10:15:33.772+00:00 | main | INFO  | StorageSubCommand$RevertVariablesStorage | Reverted variable storage for key chainHeadHash
2023-06-06 10:15:33.777+00:00 | main | INFO  | StorageSubCommand$RevertVariablesStorage | Reverted variable storage for key finalizedBlockHash
2023-06-06 10:15:33.777+00:00 | main | INFO  | StorageSubCommand$RevertVariablesStorage | Reverted variable storage for key safeBlockHash
2023-06-06 10:15:33.785+00:00 | main | INFO  | StorageSubCommand$RevertVariablesStorage | Reverted variable storage for key forkHeads
2023-06-06 10:15:33.786+00:00 | main | INFO  | StorageSubCommand$RevertVariablesStorage | Reverted variable storage for key local-enr-seqno
```

## Fixed Issue(s)

prerequisite for #5475

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/storage/keyvalue/KeyValueStoragePrefixedKeyBlockchainStorage.java [hunk 1]
```diff
@@ -26,64 +33,58 @@
...
-  private static final Bytes FORK_HEADS_KEY =
-      Bytes.wrap("forkHeads".getBytes(StandardCharsets.UTF_8));
-  private static final Bytes FINALIZED_BLOCK_HASH_KEY =
-      Bytes.wrap("finalizedBlockHash".getBytes(StandardCharsets.UTF_8));
-  private static final Bytes SAFE_BLOCK_HASH_KEY =
-      Bytes.wrap("safeBlockHash".getBytes(StandardCharsets.UTF_8));
-
+  private static final Logger LOG =
+      LoggerFactory.getLogger(KeyValueStoragePrefixedKeyBlockchainStorage.class);
   private static final Bytes VARIABLES_PREFIX = Bytes.of(1);
```
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/storage/keyvalue/KeyValueStoragePrefixedKeyBlockchainStorage.java:49]: perhaps rename to DEPRECATED_VARIABLES_PREFIX

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/storage/keyvalue/VariablesKeyValueStorage.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+  private Hash bytesToHash(final Bytes bytes) {
+    return Hash.wrap(Bytes32.wrap(bytes, 0));
+  }
+
+  Optional<Bytes> getVariable(final Keys key) {
+    return variables.get(key.toByteArray()).map(Bytes::wrap);
+  }
+
+  public static class Updater implements VariablesStorage.Updater {
```
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/storage/keyvalue/VariablesKeyValueStorage.java:83]: Was there a need to have a separate updater apart from the blockchain storage updater?

**Review by garyschulte — APPROVED** [review]
Solid :+1: minor feedback

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/storage/keyvalue/KeyValueStoragePrefixedKeyBlockchainStorage.java (hunk 1 above)
Comment (fab-10) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/storage/keyvalue/KeyValueStoragePrefixedKeyBlockchainStorage.java:49]: added the `Deprecated` annotation

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/storage/keyvalue/VariablesKeyValueStorage.java (hunk 2 above)
Comment (fab-10) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/storage/keyvalue/VariablesKeyValueStorage.java:83]: it is possible to use the variables storage outside of the blockchain, [like here](https://github.com/fab-10/besu/blob/85b0aebd71c1960e94730274b0118b961e8216a9/ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/PeerDiscoveryAgent.java#L230), and will also be use to store the world state variables in future

**Pull request merged**